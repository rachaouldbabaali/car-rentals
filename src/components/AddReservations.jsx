import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { v4 as uuidv4 } from 'uuid';
import { createReservation } from '../redux/reservations/reservationSlice';
import { getUser } from '../util/auth';
import { fetchCars } from '../redux/cars/carsSlice';

function AddReservations() {
  const [state, setState] = useState({
    user: getUser.id,
    carId: '',
    city: '',
    rentDate: '',
    returnDate: '',
  });

  const { cars } = useSelector((store) => store.cars);

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCars());
  }, [dispatch]);

  const handleChange = (e) => {
    setState({
      ...state,
      [e.target.name]: e.target.value,
    });
  };

  const cities = [
    'London',
    'New York City',
    'Dar es Salaam',
    'Kigali',
    'Lagos',
  ];

  const cityOptions = cities.map((city) => (
    <option value={city} key={uuidv4()}>
      {city}
    </option>
  ));

  const carOptions = Array.isArray(cars)
    ? cars.map((car) => (
      <option value={car.id} key={car.id}>
        {car.name}
      </option>
    ))
    : null;

  const handleFormSubmit = (e) => {
    e.preventDefault();
    dispatch(
      createReservation(state),
      setState({
        username: '',
        carBrand: '',
        city: '',
        rentDate: '',
        returnDate: '',
      }),
    );
  };

  return (
    <section className="bg-[#95BF02] h-full m-0 md:fixed">
      <div className="p-5 text-justify md:mx-auto md:w-1/2 md:my-[10%]">
        <h2 className="text-center">RESERVE A CAR TEST-RIDE</h2>
        <p className="my-2">
          There are 34 different versions of the Vespa. Today five series are in
          production: the classic manual transmission PX and the modern CVT
          transmission S: LX. GT, and GTS: We have showrooms all over the globe
          which some include test-riding facilities. if you wish to find out if
          a test-ride is available in your area, please use the selector below.
          London Book Now
        </p>

        <form onSubmit={handleFormSubmit}>
          <label htmlFor="input-name">
            <input
              type="text"
              id="input-name"
              name="username"
              value={state.username}
              onChange={handleChange}
              placeholder="Username"
            />
          </label>

          <label htmlFor="input-car">
            <select
              name="city"
              id="input-car"
              value={state.city}
              onChange={handleChange}
            >
              <option value="" disabled>
                --Pick a city --
              </option>
              {cityOptions}
            </select>
          </label>
          <label htmlFor="input-car">
            <select
              name="carId"
              id="input-car"
              value={state.carBrand}
              onChange={handleChange}
            >
              <option value="" disabled>
                --Pick a Car brand --
              </option>
              {carOptions}
            </select>
          </label>
          <label htmlFor="input-date">
            <input
              type="date"
              id="input-date"
              name="rentDate"
              value={state.rentDate}
              onChange={handleChange}
              placeholder="date"
            />
          </label>
          <label htmlFor="input-date">
            <input
              type="date"
              id="input-date"
              name="returnDate"
              value={state.returnDate}
              onChange={handleChange}
            />
          </label>
          <button type="submit">RESERVE NOW</button>
        </form>
      </div>
    </section>
  );
}

export default AddReservations;
