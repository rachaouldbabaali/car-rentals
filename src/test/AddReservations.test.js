import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import axios from 'axios';
import renderer from 'react-test-renderer';
import myReservations from './reservationsData';
import AddReservations from '../components/AddReservations';
import myCars from './carsData';

jest.mock('axios');

const reducer = (state = { reservations: { reservations: myReservations }, cars: { cars: myCars} }) => state;

const store = configureStore({ reducer });

describe('Should render correctly', () => {
  it('It should add Reservations', async () => {
    axios.get.mockResolvedValue({ data: myReservations });
    const tree = renderer
      .create(
        <Provider store={store}>
          <AddReservations />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
