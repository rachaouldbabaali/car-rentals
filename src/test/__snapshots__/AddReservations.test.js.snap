// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly It should add Reservations 1`] = `
<section
  className="bg-[#95BF02] fixed top-0 bottom-0 left-0 right-0 text-white"
>
  <div
    className="p-5 text-justify md:mx-auto md:w-1/2 md:my-[10%]"
  >
    <h2
      className="text-center font-bold"
    >
      RESERVE A CAR TEST-RIDE
    </h2>
    <p
      className="my-2 text-[9px] md:text-center"
    >
      There are 34 different versions of the Vespa. Today five series are in production: the classic manual transmission PX and the modern CVT transmission S: LX. GT, and GTS: We have showrooms all over the globe which some include test-riding facilities. if you wish to find out if a test-ride is available in your area, please use the selector below. London Book Now
    </p>
    <form
      className="md:grid grid-cols-2 gap-3"
      onSubmit={[Function]}
    >
      <input
        className="border rounded-3xl bg-[#95BF02] hover:bg-white hover:text-[#95BF02]"
        id="input-car"
        name="destination"
        onChange={[Function]}
        placeholder="City"
        type="text"
        value=""
      />
      <select
        className="border rounded-3xl bg-[#95BF02] hover:bg-white hover:text-[#95BF02]"
        id="input-car"
        name="car_id"
        onChange={[Function]}
        value=""
      >
        <option
          disabled={true}
          value=""
        >
          --Pick a Car brand --
        </option>
        <option
          value={1}
        >
          Car 1
        </option>
        <option
          value={2}
        >
          Car 2
        </option>
        <option
          value={3}
        >
          Car 3
        </option>
        <option
          value={4}
        >
          Car 4
        </option>
        <option
          value={5}
        >
          Car 5
        </option>
        <option
          value={6}
        >
          Car 6
        </option>
      </select>
      <label
        htmlFor="start_date"
      >
        Start:
        <input
          className="border rounded-3xl bg-[#95BF02] hover:bg-white hover:text-[#95BF02] w-full"
          id="input-date"
          name="rental_date"
          onChange={[Function]}
          placeholder="date time"
          type="date"
          value=""
        />
      </label>
      <label
        htmlFor="start_date"
      >
        End:
        <input
          className="border rounded-3xl bg-[#95BF02] hover:bg-white hover:text-[#95BF02] w-full"
          id="input-date"
          name="date_return"
          onChange={[Function]}
          type="date"
          value=""
        />
      </label>
      <button
        className="col-span-2 border rounded-3xl bg-[#95BF02] hover:bg-white hover:text-[#95BF02] w-1/2 mx-auto"
        type="submit"
      >
        RESERVE NOW
      </button>
    </form>
  </div>
</section>
`;

exports[`Should render correctly It should get all mission loaded 1`] = `
<section
  className="bg-[#95BF02] fixed top-0 bottom-0 left-0 right-0 text-white"
>
  <div
    className="p-5 text-justify md:mx-auto md:w-1/2 md:my-[10%]"
  >
    <h2
      className="text-center font-bold"
    >
      RESERVE A CAR TEST-RIDE
    </h2>
    <p
      className="my-2 text-[9px] md:text-center"
    >
      There are 34 different versions of the Vespa. Today five series are in production: the classic manual transmission PX and the modern CVT transmission S: LX. GT, and GTS: We have showrooms all over the globe which some include test-riding facilities. if you wish to find out if a test-ride is available in your area, please use the selector below. London Book Now
    </p>
    <form
      className="md:grid grid-cols-2 gap-3"
      onSubmit={[Function]}
    >
      <input
        className="border rounded-3xl bg-[#95BF02] hover:bg-white hover:text-[#95BF02]"
        id="input-car"
        name="destination"
        onChange={[Function]}
        placeholder="City"
        type="text"
        value=""
      />
      <select
        className="border rounded-3xl bg-[#95BF02] hover:bg-white hover:text-[#95BF02]"
        id="input-car"
        name="car_id"
        onChange={[Function]}
        value=""
      >
        <option
          disabled={true}
          value=""
        >
          --Pick a Car brand --
        </option>
        <option
          value={1}
        >
          Car 1
        </option>
        <option
          value={2}
        >
          Car 2
        </option>
        <option
          value={3}
        >
          Car 3
        </option>
        <option
          value={4}
        >
          Car 4
        </option>
        <option
          value={5}
        >
          Car 5
        </option>
        <option
          value={6}
        >
          Car 6
        </option>
      </select>
      <label
        htmlFor="start_date"
      >
        Start:
        <input
          className="border rounded-3xl bg-[#95BF02] hover:bg-white hover:text-[#95BF02] w-full"
          id="input-date"
          name="rental_date"
          onChange={[Function]}
          placeholder="date time"
          type="date"
          value=""
        />
      </label>
      <label
        htmlFor="start_date"
      >
        End:
        <input
          className="border rounded-3xl bg-[#95BF02] hover:bg-white hover:text-[#95BF02] w-full"
          id="input-date"
          name="date_return"
          onChange={[Function]}
          type="date"
          value=""
        />
      </label>
      <button
        className="col-span-2 border rounded-3xl bg-[#95BF02] hover:bg-white hover:text-[#95BF02] w-1/2 mx-auto"
        type="submit"
      >
        RESERVE NOW
      </button>
    </form>
  </div>
</section>
`;
